Store Orders Page: Specific Order UI

alexzhang1618 opened the pull request:
# Description

What changes did you make? List all distinct problems that this PR addresses. Explain any relevant
motivation or context.
- Implemented examining the info of a specific order

Things I still need to do:
- [x] Discount price displays
- [ ] Flows for doing stuff (cancelling, rescheduling, etc etc)
- [ ] animation for opening a card? this might be difficult
- [x] The specific event card header orientation for mobile (the status symbol moves to the top)

## Changes

- [Fill in here]

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) <!-- Edit this type of change if you select this -->

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [ ] locally on mobile - use https://ngrok.io to get a copy on a mobile device
- [ ] on the live deployment preview on Desktop.
- [ ] on the live deployment preview on Mobile.
- [ ] I have run and passed all new and existing Cypress tests. Add screenshots below.

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented my code's `src/lib` functions and commented hard to understand areas
      anywhere else.
- [x] My changes produce no new warnings.

# Screenshots
<img width="1440" alt="Screenshot 2023-12-30 at 12 05 36 AM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/e98fc4a8-3fe5-454c-bbd6-23e1a7274411">

<img width="356" alt="Screenshot 2023-12-30 at 12 06 00 AM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/ca0a19c6-66f0-4783-bed4-c6fa4161c502">

alexzhang1618 committed:
UI for examining a specific order

SheepTester commented on src/components/store/OrderCard/index.tsx:
```diff
@@ -28,28 +31,38 @@ export const orderStatusColor: { [_ in OrderStatus]: string } = {
 
 interface OrderCardProps {
   order: PublicOrder;
+  focusedOrder?: PublicOrderWithItems;
```
hmm, why force at most one order card to be open at a time? [exclusive accordions hurt accessibility and usability](https://github.com/openui/open-ui/issues/925), and it'd also perhaps be easier just for each `OrderCard` to keep track of their own state

alexzhang1618 replied:
Done, I'm writing this with `useEffect` and `useState` so we can stagger the number of getOrder API calls to be on demand instead of all on page load.

SheepTester commented on src/components/store/OrderSummary/index.tsx:
```diff
@@ -0,0 +1,72 @@
+    }
+  });
+
+  const updatedItems = Array.from(itemMap, ([, value]) => value);
```
you can also do

```suggestion
  const updatedItems = Array.from(itemMap.values());
```

SheepTester commented on src/components/store/OrderSummary/index.tsx:
```diff
@@ -0,0 +1,72 @@
+    if (existingItem) {
+      existingItem.quantity += 1;
+
+      itemMap.set(existingItem.option.uuid, existingItem);
```
I don't think this line is necessary since you're mutating the original object

SheepTester commented on src/components/store/OrderSummary/index.tsx:
```diff
@@ -0,0 +1,72 @@
+            </div>
+            {item.option.metadata && (
+              <div className={styles.label}>
+                <Typography variant="h5/bold">{`${capitalize(
+                  item.option.metadata.type
+                )}: `}</Typography>
```
You could also let CSS do this for you, using `text-transform: capitalize;`, though it capitalizes every word instead of just the first letter

alexzhang1618 replied:
https://stackoverflow.com/questions/52165435/why-css-text-transform-capitalize-not-working

sadly, it doesn't decapitalize the rest of the word 😢 
It's still possible to do this thru CSS, but double cascading lowercase and capitalize seems less readable, so I think I'll keep it as is.

SheepTester reviewed (commented):
I think the text contrast could be improved. Probably the background color should be dark blue for dark theme

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/5e42592c-8f70-4470-9789-145aa5019ffa)

maybe on mobile, the text should show below the image?

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/e5db1df2-6b09-489f-9282-536e5c7ed3e0)

> * animation for opening a card? this might be difficult

yes it will be, i recommend putting this off for later

alexzhang1618 committed:
Move API handling for individual orders into components instead of big page

alexzhang1618 committed:
Implement discount diamonds component

alexzhang1618 committed:
Tiny cleanup

alexzhang1618 committed:
Fix visuals and mobile layouts

alexzhang1618 committed:
Tiny bugfix, testing with different statuses

alexzhang1618 committed:
Add action buttons for orders

alexzhang1618 committed:
Add view event details

alexzhang1618 commented:
Since the last review: 

- Fixed some color contrast stuff for order status
- Implemented the Discount price views
- Made each card openable and moved API code into the card
- Added mobile layout for status
- Added action buttons to view pickup event, reschedule, and cancel order
  - I'm going to wait until Modal is merged in #91 before implementing these since I want to use it for these buttons

<img width="1440" alt="Screenshot 2023-12-30 at 1 37 38 PM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/a5a35f05-f1e4-4eae-bc5a-8b9dc74ca3ca">
<img width="1440" alt="Screenshot 2023-12-30 at 1 38 20 PM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/349149de-2600-4f7d-9609-31110673c0f2">
<img width="1440" alt="Screenshot 2023-12-30 at 1 38 29 PM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/78ae8548-9e8a-404d-aae9-ea8edeffb249">
<img width="369" alt="Screenshot 2023-12-30 at 1 38 56 PM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/ad8c5dca-08cd-45c9-8425-94647dfe1995">
<img width="369" alt="Screenshot 2023-12-30 at 1 41 19 PM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/2f5fb941-6ab6-4311-9ba1-08997268082c">

alexzhang1618 committed:
Small CSS fixes for action buttons

farisashai reviewed (changes requested):
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/33165426/56d8ec72-7973-4366-98c0-6c21ea937821)

can you fix the blue color on the orders card and dropdown and also the padding for the store navbar is inconsistent with the page padding

farisashai commented on src/components/store/Diamonds/index.tsx:
```diff
@@ -5,14 +5,26 @@ const compactFormat = new Intl.NumberFormat('en-US', { notation: 'compact' });
+  const discountActive = discount !== undefined && discount < count;
+  return (
+    <span className={className}>
+      <span style={discountActive ? { textDecoration: 'line-through' } : {}}>
```
probably cleaner for the future to have a discount active classmate rather than doing all of this online styling

farisashai commented on src/components/store/Diamonds/index.tsx:
```diff
@@ -5,14 +5,26 @@ const compactFormat = new Intl.NumberFormat('en-US', { notation: 'compact' });
+      </span>
+      {discountActive && (
+        <span style={{ color: 'var(--theme-danger-1)' }}>
+          {' '}
```
```suggestion
          &nbsp;
```

farisashai commented on src/components/store/OrderCard/index.tsx:
```diff
@@ -28,28 +32,61 @@ export const orderStatusColor: { [_ in OrderStatus]: string } = {
+  const orderOpen = open && orderData !== null;
+
+  useEffect(() => {
+    if (open && orderData === null) {
```
This looks like the same Boolean you defined on 41

alexzhang1618 replied:
It's not quite the same, this makes sure we don't refetch if we've already done so `orderData === null` versus `orderData !== null`

farisashai commented on src/components/store/OrderCard/style.module.scss:
```diff
@@ -1,70 +1,113 @@
-      border: 1px solid vars.$disabled;
-      color: vars.$disabled;
+    &.focused {
+      border-bottom-left-radius: 0;
```
```suggestion
      border-radius: 0.5rem 0.5rem 0 0;
```

alexzhang1618 committed:
Fix comments

alexzhang1618 commented:
fixed by adding to reset.scss, I also removed the change in EventCard since this covers it as well. padding also fixed too

alexzhang1618 committed:
Address comments
